cassiopeia/solver/missing_data_methods.py

```diff
@@ -2,74 +2,87 @@
 
     Returns:
         A tuple of lists, representing the left and right partitions with
         missing samples imputed
     """
-    subset_character_matrix_left = character_matrix[left_set, :]
-    subset_character_matrix_right = character_matrix[right_set, :]
 
-    for i in missing:
-        left_score = 0
-        right_score = 0
+    # A helper function to calculate the number of shared character/state pairs
+    # shared between a missing sample and a side of the partition
+    indices = list(character_matrix.index)
+    character_array = character_matrix.to_numpy()
+    left_indices = solver_utilities.convert_sample_names_to_indices(
+        indices, left_set
+    )
+    right_indices = solver_utilities.convert_sample_names_to_indices(
+        indices, right_set
+    )
+    missing_indices = solver_utilities.convert_sample_names_to_indices(
+        indices, missing
+    )
 
+    def score_side(subset_character_matrix, missing_sample):
+        score = 0
         for char in range(character_matrix.shape[1]):
-            state = character_matrix[i][char]
+            state = character_array[missing_sample, char]
             if state != missing_char and state != 0:
                 state_counts = np.unique(
-                    subset_character_matrix_left[:, char], return_counts=True
+                    subset_character_matrix[:, char], return_counts=True
                 )
                 ind = np.where(state_counts[0] == state)
                 if len(ind[0]) > 0:
-                    left_score += state_counts[1][ind[0][0]]
-                else:
-                    left_score += 0
+                    if weights:
+                        score += (
+                            weights[char][state] * state_counts[1][ind[0][0]]
+                        )
+                    else:
+                        score += state_counts[1][ind[0][0]]
+        return score
 
-        for char in range(character_matrix.shape[1]):
-            state = character_matrix[i][char]
-            if state != missing_char and state != 0:
-                state_counts = np.unique(
-                    subset_character_matrix_right[:, char], return_counts=True
-                )
-                ind = np.where(state_counts[0] == state)
-                if len(ind[0]) > 0:
-                    right_score += state_counts[1][ind[0][0]]
-                else:
-                    right_score += 0
+    subset_character_array_left = character_array[left_indices, :]
+    subset_character_array_right = character_array[right_indices, :]
+
+    for sample in missing_indices:
+        left_score = score_side(subset_character_array_left, sample)
+        right_score = score_side(subset_character_array_right, sample)
 
         if left_score / len(left_set) > right_score / len(right_set):
-            left_set.append(i)
+            left_set.append(indices[sample])
         else:
-            right_set.append(i)
+            right_set.append(indices[sample])
```

Comment:
Totally fine logic, but a nitpicky comment -- you seem to use the terminology that "sample" is a name of a row in the original character matrix and index is the integer representation of this. But here, these nomenclatures are swapped, I believe. It makes the code a bit less readable (I would have imagined it would make more sense to say `left_set.append(samples[index])`, for example. Does that make sense?)

Reply:
I agree here I'll fix this up!